x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/inference/RerankOperator.java

```diff
@@ -0,0 +1,198 @@
+    }
+
+    private Page buildOutput(Page inputPage, InferenceAction.Response inferenceResponse) {
+        if (inferenceResponse.getResults() instanceof RankedDocsResults rankedDocsResults) {
+            return buildOutput(inputPage, rankedDocsResults);
+
+        }
+
+        throw new IllegalStateException(
+            "Inference result has wrong type. Got ["
+                + inferenceResponse.getResults().getClass()
+                + "] while expecting ["
+                + RankedDocsResults.class
+                + "]"
+        );
+    }
+
+    private Page buildOutput(Page inputPage, RankedDocsResults rankedDocsResults) {
+        int blockCount = Integer.max(inputPage.getBlockCount(), scoreChannel + 1);
+        Block[] blocks = new Block[blockCount];
+
+        try {
+            for (int b = 0; b < blockCount; b++) {
+                if (b == scoreChannel) {
+                    blocks[b] = buildScoreBlock(inputPage, rankedDocsResults);
+                } else {
+                    blocks[b] = inputPage.getBlock(b);
+                    blocks[b].incRef();
+                }
+            }
+            return new Page(blocks);
+        } catch (Exception e) {
+            Releasables.closeExpectNoException(blocks);
+            throw (e);
+        }
+    }
+
+    private Block buildScoreBlock(Page inputPage, RankedDocsResults rankedDocsResults) {
+        Double[] sortedRankedDocsScores = new Double[inputPage.getPositionCount()];
+
+        try (DoubleBlock.Builder scoreBlockFactory = blockFactory.newDoubleBlockBuilder(inputPage.getPositionCount())) {
+            for (RankedDocsResults.RankedDoc rankedDoc : rankedDocsResults.getRankedDocs()) {
+                sortedRankedDocsScores[rankedDoc.index()] = (double) rankedDoc.relevanceScore();
+            }
+
+            for (int pos = 0; pos < inputPage.getPositionCount(); pos++) {
+                if (sortedRankedDocsScores[pos] != null) {
+                    scoreBlockFactory.appendDouble(sortedRankedDocsScores[pos]);
+                } else {
+                    scoreBlockFactory.appendNull();
+                }
+            }
+
+            return scoreBlockFactory.build();
+        }
+    }
+
+    private InferenceAction.Request buildInferenceRequest(Page inputPage) {
+        try (BytesRefBlock encodedRowsBlock = (BytesRefBlock) rowEncoder.eval(inputPage)) {
+            assert (encodedRowsBlock.getPositionCount() == inputPage.getPositionCount());
+            String[] inputs = new String[inputPage.getPositionCount()];
+            BytesRef buffer = new BytesRef();
+
+            for (int pos = 0; pos < inputPage.getPositionCount(); pos++) {
+                if (encodedRowsBlock.isNull(pos)) {
+                    inputs[pos] = "";
+                } else {
+                    buffer = encodedRowsBlock.getBytesRef(encodedRowsBlock.getFirstValueIndex(pos), buffer);
```

Comment:
Not for this round but it would be nice to be able to limit the extra memory here. Currently we have to run a full page and duplicate the entire fields content to run the inference request.

Reply:
Added a follow-up in the meta-issue.